docs/DesignDocs/Attestation_API_Proposal.md

```diff
@@ -685,6 +685,109 @@ since legacy applications don't call these API functions explicitly.
  `oe_attester_shutdown()` and `oe_verifier_shutdown()` are NOP, so there
  is no need to call them upon application close.
 
+## Proposed V0.11 Implementation
+
+### Goals
+1. Consolidate SGX_ECDSA_P256 format IDs.
+2. Deprecate oe_report definitions.
+
+### Motivation
+The 2 extra SGX P256 evidence formats defined in V0.10 implementation can be reduced/deprecated to make the API surface cleaner:
+- `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT`
+- `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE`
+
+For every listed format, `oe_get_evidence()` can wrap with
+an attestation header when the `OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID` bit in its `flags` parameter is set. These are all possible SGX_ECDSA_P256 format structures that can be generated by oe_get_evidence():
+
+| Format ID | Evidence structure  |
+| -- | - |
+| 1. `OE_FORMAT_UUID_SGX_ECDSA_P256` | `oe_report_header || SGX_ECDSA_quote(hash) || custom_claims` |
+| 2. `OE_FORMAT_UUID_SGX_ECDSA_P256 + flag` | `oe_attestation_header || oe_report_header || SGX_ECDSA_quote(hash) || custom_claims` |
+| 3. `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT` | `oe_report_header || SGX_ECDSA_quote(custom_claims)` |
+| 4. `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT + flag` | `oe_attestation_header || oe_report_header || SGX_ECDSA_quote(custom_claims)` |
+| 5. `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE` | `SGX_ECDSA_quote(custom_claims)` |
+| 6. `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE + flag` | `oe_attestation_header || SGX_ECDSA_quote(custom_claims)` |
+
+For all evidence structures above, the hearts of the evidences are the same - SGX ECDSA_P256 quote. SGX ECDSA_P256 quote should to be like any other evidence format which is presented in 2 formats - with or without embedded format id. For example, format XYZ is presented as:
+1. Without flag - `Evidence_XYZ`
+2. With flag - `oe_attestation_header || Evidence_XYZ || custom_claims`
+
+That is, we only need to support these evidence structures:
+- Format 2: Generated by previous version oe_get_evidence().
+- Format 3: Generated by legacy oe_get_report() which will eventually be deprecated.
+- Format 5: Generated by SGX SDK.
+
+### User Experience
+
+#### Attester
+- To generate format 2:
+```
+oe_get_evidence(
+    OE_FORMAT_UUID_SGX_ECDSA_P256,
+    OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID,
+    custom_claims,
+    size_t custom_claims_size,
+    ...);
+```
+
+- To generate format 3:
+```
+oe_get_evidence(
+    OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT,
+    0,  // Must be 0
+    custom_claims,
+    size_t custom_claims_size,  // Must be <= 64 bytes
+    ...);
+```
+
+- To generate format 5:
+```
+oe_get_evidence(
+    OE_FORMAT_UUID_SGX_ECDSA_P256,
+    0,
```

Comment:
With this proposal, the OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID flag is effectively extended / redefined so that the OE SDK framework not just wraps the plugin evidence with a header containing format ID, but also hashes the custom claims and passes the hash to the plugin. Missing of this flag requires a plugin to embed the custom claims in plugin-specific way (e.g. embedding it in  the 64-byte SGX report data field).

There has been discussions in SIG-attestation about some TEEs not having the custom claims size limit in the SGX report data field (64 bytes). For those plugins, forcing them to take a hash may not be the optimal solution.

Reply:
I'd go for consistency. If OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID is set, custom claims is always appended to the evidence and report_data field contains hash value. This simplifies the code.

In the future when report_data field allows more than typical custom_claims size needed, we can make adjustment. From what I've read, 64 bytes report_data seems common.

Reply:
We could define another flag to control the hashing behavior, but if we did that it should be done now instead of later.   My hope though is that TEEs other than SGX would never want to use EMBED_FORMAT_ID anyway and would instead use a standard format (for which EMBED_FORMAT_ID would break interop), so I'm hoping that over time EMBED_FORMAT_ID becomes a legacy thing.

Reply:
A new flag sounds like a good option. On the other hand, all flags are effectively part of the format ID, and will need to be communicated / agreed upon between attesters and verifiers, so the verifiers know how to process. So it's preferable to minimize or eliminate flags.

In the current implementation, `oe_verify_evidence()` can output either claim `OE_CLAIM_SGX_REPORT_DATA`, or claim `OE_CLAIM_CUSTOM_CLAIMS`, or neither. In the proposal, what claims will be output for each format?

Reply:
An alternative to the proposal could be for `oe_get_evidence()` to support `OE_FORMAT_UUID_SGX_ECDSA_P256` and `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE`, and for `oe_verify_evidence()` to support all 3 format IDs. There is no change in the current API definition and implmeentation, except for removing support of `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT` in the SGX ECDSA attester plugin library.

`oe_verify_evidence() support of all 3 format IDs is needed for backward compatibility, to meet Larry’s needs.

Support of generation of SGX ECDSA evidence in format 5 is needed for backward compatibility, to help porting of existing SGX SDK solutions to OE SDK, since those applications already expects raw SGX quotes.

Support of generation of SGX ECDSA evidence in formats 3 and 5 was added to also allow testing of `oe_verify_evidence()` for those formats. But as implemented in `tests/attestation_plugin/plugin/tests.c` lines 786 – 843, for testing purpose, those evidence can be constructed from evidence in format 1 or 2. So support of formats 3 is no longer needed.

As the need for backward compatibility reduces, `oe_get_evicence() support of format 3 and 4, and `oe_verify_evidence()` support of formats 3 to 6 can be phased out when possible.

The API documentation can be made to state clearly that applications should use `oe_get_evidence()`  only with `OE_FORMAT_UUID_SGX_ECDSA_P256` with `OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID` unless there is a need for backward compatibility.

Reply:
> A new flag sounds like a good option. On the other hand, all flags are effectively part of the format ID, and will need to be communicated / agreed upon between attesters and verifiers, so the verifiers know how to process. So it's preferable to minimize or eliminate flags.
> 
> In the current implementation, `oe_verify_evidence()` can output either claim `OE_CLAIM_SGX_REPORT_DATA`, or claim `OE_CLAIM_CUSTOM_CLAIMS`, or neither. In the proposal, what claims will be output for each format?

Format 2 `oe_verify_evidence(NULL)` will contain `OE_CLAIM_CUSTOM_CLAIMS`. It's the same as current implementation.
Format 3 `oe_verify_evidence(OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT)` and 5 `oe_verify_evidence(OE_FORMAT_UUID_SGX_ECDSA_P256)` will contain `OE_CLAIM_SGX_REPORT_DATA`. Format 5 is basically trying to replace current `oe_verify_evidence(OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE)`. And we hope to deprecate the format 3 eventually.

Reply:
The proposal is to keep the public API simple and remove redundancy. Adding another flag doubles the evidence formats. I agree with @shnwc that it's preferable to minimize or eliminate flags.

Suggest rename OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID to something that reflects the fact.

To make the responsibility clear, `OE_CLAIM_SGX_REPORT_DATA` is to be added to the claims by SGX plugins; `OE_CLAIM_CUSTOM_CLAIMS` is added by OE SDK. In MAA's scenario, they get report_data from `OE_CLAIM_SGX_REPORT_DATA`.

Support selective formats is an alternative but it does not present clean and concise API. Many people don't read document.

Reply:
> Support selective formats is an alternative but it does not present clean and concise API. Many people don't read document.

Personally I think the current API definition is clean. In addition to clarification in API documentation, we could put the definitions of format IDs etc. for backward compatibility to dedicated header file(s). Those applications that want to use the definitions for backward compatibility would need to explicitly include these header files.

Reply:
> To make the responsibility clear, OE_CLAIM_SGX_REPORT_DATA is to be added to the claims by SGX plugins; OE_CLAIM_CUSTOM_CLAIMS is added by OE SDK. In MAA's scenario, they get report_data from OE_CLAIM_SGX_REPORT_DATA

With this design, `oe_verify_evidence()` for some SGX evidence formats return both  `OE_CLAIM_SGX_REPORT_DATA` and `OE_CLAIM_CUSTOM_CLAIMS`.